### PR TITLE
[MM-16517] Migrate "Status.Get" to Sync by default

### PR DESCRIPTION
**Pull request by lpadgett**
#### Summary
This change will migrate Status.Get to sync by default so it will only use async when needed.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/11342

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -81,9 +81,9 @@ require (
 	golang.org/x/crypto v0.0.0-20190426145343-a29dc8fdc734
 	golang.org/x/image v0.0.0-20190321063152-3fc05d484e9f
 	golang.org/x/net v0.0.0-20190424112056-4829fb13d2c6
-	golang.org/x/sync v0.0.0-20190423024810-112230192c58 // indirect
```
Comment (jespino) [inline go.mod]: Please, remove this changes from your PR.

File: go.sum [hunk 2]
```diff
@@ -445,6 +445,8 @@ golang.org/x/tools v0.0.0-20190114222345-bf090417da8b/go.mod h1:n7NCudcB/nEzxVGm
 golang.org/x/tools v0.0.0-20190226205152-f727befe758c/go.mod h1:9Yl7xja0Znq3iFh3HoIrodX9oNMXvdceNzlUR8zjMvY=
 golang.org/x/tools v0.0.0-20190312170243-e65039ee4138/go.mod h1:LCzVGOaR6xXOjkQ3onu1FJEFr0SW1gC7cKk1uF8kGRs=
 golang.org/x/tools v0.0.0-20190328211700-ab21143f2384/go.mod h1:LCzVGOaR6xXOjkQ3onu1FJEFr0SW1gC7cKk1uF8kGRs=
+golang.org/x/tools v0.0.0-20190621195816-6e04913cbbac h1:MQEvx39qSf8vyrx3XRaOe+j1UDIzKwkYOVObRgGPVqI=
```
Comment (jespino) [inline go.sum]: Please, remove this changes from your PR.

File: store/sqlstore/status_store.go [hunk 3]
```diff
@@ -55,26 +55,24 @@ func (s SqlStatusStore) SaveOrUpdate(status *model.Status) store.StoreChannel {
...
-	return store.Do(func(result *store.StoreResult) {
-		var status model.Status
+func (s SqlStatusStore) Get(userId string) (*model.Status, *model.AppError) {
+	var status model.Status
 
-		if err := s.GetReplica().SelectOne(&status,
-			`SELECT
+	if err := s.GetReplica().SelectOne(&status,
+		`SELECT
 				*
```
Comment (jespino) [inline store/sqlstore/status_store.go]: Please, keep the indentation in the query consistent as was before.

File: store/sqlstore/status_store.go [hunk 4]
```diff
@@ -55,26 +55,24 @@ func (s SqlStatusStore) SaveOrUpdate(status *model.Status) store.StoreChannel {
...
 			WHERE
 				UserId = :UserId`, map[string]interface{}{"UserId": userId}); err != nil {
-			if err == sql.ErrNoRows {
-				result.Err = model.NewAppError("SqlStatusStore.Get", MISSING_STATUS_ERROR, nil, err.Error(), http.StatusNotFound)
-			} else {
-				result.Err = model.NewAppError("SqlStatusStore.Get", "store.sql_status.get.app_error", nil, err.Error(), http.StatusInternalServerError)
-			}
+		if err == sql.ErrNoRows {
+			return nil, model.NewAppError("SqlStatusStore.Get", MISSING_STATUS_ERROR, nil, err.Error(), http.StatusNotFound)
 		} else {
```
Comment (jespino) [inline store/sqlstore/status_store.go]: You don't need the else, because the if always `return`, just put the content one level up and remove the else.

File: store/sqlstore/status_store.go [hunk 5]
```diff
@@ -55,26 +55,24 @@ func (s SqlStatusStore) SaveOrUpdate(status *model.Status) store.StoreChannel {
...
-			result.Data = &status
+			return nil, model.NewAppError("SqlStatusStore.Get", "store.sql_status.get.app_error", nil, err.Error(), http.StatusInternalServerError)
 		}
-	})
+	} else {
```
Comment (jespino) [inline store/sqlstore/status_store.go]: same here

File: store/storetest/status_store.go [hunk 6]
```diff
@@ -33,7 +33,7 @@ func testStatusStore(t *testing.T, ss store.Store) {
 		t.Fatal(err)
 	}
 
-	if err := (<-ss.Status().Get(status.UserId)).Err; err != nil {
+	if _, err := ss.Status().Get(status.UserId); err != nil { //Fix
```
Comment (jespino) [inline store/storetest/status_store.go]: Why the `//Fix` comment?

File: store/storetest/status_store.go [hunk 7]
```diff
@@ -82,11 +82,11 @@ func testStatusStore(t *testing.T, ss store.Store) {
 		t.Fatal(err)
 	}
 
-	if result := <-ss.Status().Get(status.UserId); result.Err != nil {
-		t.Fatal(result.Err)
+	if _, err := ss.Status().Get(status.UserId); err != nil {
```
Comment (jespino) [inline store/storetest/status_store.go]: I would expect here something like `if status, err := ss.Status().Get(status.UserId); err != nil {` and use the returned value in the comparison later

File: store/storetest/status_store.go (hunk 6 above)
Comment (lpadgett) [reply store/storetest/status_store.go]: That was there for my own reference, it appears I forgot to take it out. My apologies

File: store/storetest/status_store.go [hunk 8]
```diff
@@ -26,7 +26,11 @@ func testStatusStore(t *testing.T, ss store.Store) {
 
 	status.LastActivityAt = 10
 
-	if err := (<-ss.Status().Get(status.UserId)).Err; err != nil {
+	if err := ss.Status().SaveOrUpdate(status); err != nil {
```
Comment (jespino) [inline store/storetest/status_store.go]: You are adding code here that wasn't there before.

File: store/storetest/status_store.go (hunk 7 above)
Comment (jespino) [reply store/storetest/status_store.go]: I miss here the `status` variable assignation.

File: store/storetest/status_store.go [hunk 9]
```diff
@@ -157,21 +160,21 @@ func testGetAllFromTeam(t *testing.T, ss store.Store) {
 	team2Member2Status := &model.Status{UserId: team2Member2.UserId, Status: model.STATUS_OFFLINE, Manual: true, LastActivityAt: model.GetMillis(), ActiveChannel: ""}
 	require.Nil(t, ss.Status().SaveOrUpdate(team2Member2Status))
 
-	if statueses, err := ss.Status().GetAllFromTeam(team1.Id); err != nil {
+	if statuses, err := ss.Status().GetAllFromTeam(team1.Id); err != nil {
```
Comment (jespino) [inline store/storetest/status_store.go]: nice catch

File: store/storetest/status_store.go (hunk 8 above)
Comment (lpadgett) [reply store/storetest/status_store.go]: The reason I did this is GoLand (the IDE I am using) alerted me that it was an invalid operation ("receive from non-chan type (*model.Status, *model.AppError)") which seemed to make sense since the StatusStore interface in store.go has (*model.Status, *model.AppError) as its return type and this portion of the test failed when run locally as well due to a similar reason ("multiple-value ss.Status().Get() in single-value context"). If this change is undesirable I will find a way to fix without altering the line mentioned here

Comment (jespino) [reply store/storetest/status_store.go]: The problem that GoLand was complaining about was because previously we had `(<-ss.Status().Get(status.UserId)).Err`, which is, read the channel from the response that comes from `Status().Get(status.UserId)` that is no longer a channel, so you have to change the code as you did a bit later, the thing is, I don't see how this: `if err := ss.Status().SaveOrUpdate(status); err != nil {` is related to the problem that you have explained.

Comment (lpadgett) [reply store/storetest/status_store.go]: Upon further review (and getting some sleep) understood your original complaint and noticed my mistake, my apologies.

Comment (jespino) [reply store/storetest/status_store.go]: No problem :)

File: store/storetest/status_store.go [hunk 10]
```diff
@@ -68,11 +68,10 @@ func testStatusStore(t *testing.T, ss store.Store) {
 		t.Fatal(err)
 	}
 
-	if result := <-ss.Status().Get(status.UserId); result.Err != nil {
-		t.Fatal(result.Err)
+	if statusParameter, err := ss.Status().Get(status3.UserId); err != nil {
```
Comment (jespino) [inline store/storetest/status_store.go]: You are changing here the variable used to get the status from `status` to `status3`

File: store/storetest/status_store.go [hunk 11]
```diff
@@ -64,15 +64,14 @@ func testStatusStore(t *testing.T, ss store.Store) {
 		}
 	}
 
-	if err := (<-ss.Status().ResetAll()).Err; err != nil {
+	if err := ss.Status().ResetAll(); err != nil {
```
Comment (jespino) [inline store/storetest/status_store.go]: You are making here an unrelated change, It shouldn't be changed because is not related to the PR, besides is not correct, because now we are checking that the `store.StoreResult` returned there is not a `nil` instead of checking the `err`.
